Test failures (regressions) with Python 3.12

Not the “Poetry way,” but a clean way to reproduce:

```
$ python3.11 -VV
Python 3.11.4 (main, Jun  7 2023, 00:00:00) [GCC 13.1.1 20230511 (Red Hat 13.1.1-2)]
```

```
gh repo clone p1c2u/jsonschema-spec
cd jsonschema-spec
python3.11 -m venv _e
. _e/bin/activate
pip install -e .
pip install pytest pytest-cov responses
python -m pytest
```

This confirms everything is working as expected with Python 3.11:

```
================================================= 27 passed in 0.29s =================================================
```

Now, repeat with Python 3.12:

```
$ python3.12 -VV
Python 3.12.0b3 (main, Jun 21 2023, 00:00:00) [GCC 13.1.1 20230614 (Red Hat 13.1.1-4)]
```

```
python3.12 -m venv _f
. _f/bin/activate
pip install -e .
pip install pytest pytest-cov responses
python -m pytest
```

```
============================================== short test summary info ===============================================
FAILED tests/integration/test_schema_path.py::TestSchemaPathFromFilePath::test_file_path - AttributeError: '_io.BufferedReader' object has no attribute 'decode'
FAILED tests/integration/test_schema_path.py::TestSchemaPathFromFilePath::test_file_path_relative - AttributeError: '_io.BufferedReader' object has no attribute 'decode'
FAILED tests/integration/test_schema_path.py::TestSchemaPathOpen::test_file - referencing.exceptions.Unresolvable: /tmp/tmpvferprav#/Info
FAILED tests/integration/test_schema_path.py::TestSchemaPathOpen::test_file_ref - AttributeError: '_io.BufferedReader' object has no attribute 'decode'
FAILED tests/integration/test_schema_path.py::TestSchemaPathOpen::test_double_ref - AttributeError: '_io.BufferedReader' object has no attribute 'decode'
FAILED tests/integration/test_schema_path.py::TestSchemaPathOpen::test_ref_recursive - referencing.exceptions.Unresolvable: recursive.yaml#/parameters/RecursiveReference
FAILED tests/unit/test_paths.py::TestSchemaPathFromPath::test_no_kwargs - AttributeError: '_io.BufferedReader' object has no attribute 'decode'
FAILED tests/unit/test_paths.py::TestSchemaPathFromFilePath::test_no_kwargs - AttributeError: '_io.BufferedReader' object has no attribute 'decode'
============================================ 8 failed, 19 passed in 1.29s ============================================
```

Full output: [python312.txt](https://github.com/p1c2u/jsonschema-spec/files/12010184/python312.txt)

--- a/jsonschema_spec/handlers/file.py
+++ b/jsonschema_spec/handlers/file.py
@@ -1,11 +1,10 @@
 """JSONSchema spec handlers file module."""
-from io import StringIO
 from json import dumps
 from json import loads
-from typing import IO
 from typing import Any
-from typing import List
-from typing import Union
+from typing import ContextManager
+from typing import Optional
+from typing import Tuple
 from urllib.parse import urlparse
 
 from yaml import load
@@ -21,39 +20,51 @@ class FileHandler:
     def __init__(self, loader: Any = JsonschemaSafeLoader):
         self.loader = loader
 
-    def __call__(self, f: SupportsRead) -> Any:
-        data = self._load(f)
+    def __call__(self, stream: SupportsRead) -> Any:
+        data = self._load(stream)
         return loads(dumps(data))
 
-    def _load(self, f: SupportsRead) -> Any:
-        return load(f, self.loader)
+    def _load(self, stream: SupportsRead) -> Any:
+        return load(stream, self.loader)
 
 
-class BaseFilePathHandler(FileHandler):
+class BaseFilePathHandler:
     """Base file path handler."""
 
-    allowed_schemes: List[str] = NotImplemented
+    allowed_schemes: Tuple[str, ...] = NotImplemented
 
-    def __call__(self, uri: Union[SupportsRead, str]) -> Any:
-        if isinstance(uri, SupportsRead):
-            return super().__call__(uri)
+    def __init__(
+        self, *allowed_schemes: str, file_handler: Optional[FileHandler] = None
+    ):
+        self.allowed_schemes = allowed_schemes or self.allowed_schemes
+        self.file_handler = file_handler or FileHandler()
 
+    def __call__(self, uri: str) -> Any:
         parsed_url = urlparse(uri)
         if parsed_url.scheme not in self.allowed_schemes:
-            raise ValueError("Not allowed scheme")
+            raise ValueError(f"Scheme {parsed_url.scheme} not allowed")
 
-        return self._open(uri)
+        with self._open(uri) as stream:
+            return self.file_handler(stream)
 
-    def _open(self, uri: str) -> Any:
+    def _open(self, uri: str) -> ContextManager[SupportsRead]:
         raise NotImplementedError
 
 
 class FilePathHandler(BaseFilePathHandler):
     """File path handler."""
 
-    allowed_schemes = ["file"]
+    allowed_schemes = ("file",)
 
-    def _open(self, uri: str) -> Any:
+    def __init__(
+        self,
+        *allowed_schemes: str,
+        file_handler: Optional[FileHandler] = None,
+        encoding: str = "utf-8",
+    ):
+        super().__init__(*allowed_schemes, file_handler=file_handler)
+        self.encoding = encoding
+
+    def _open(self, uri: str) -> ContextManager[SupportsRead]:
         filepath = uri_to_path(uri)
-        with open(filepath) as fh:
-            return super().__call__(fh)
+        return open(filepath, encoding=self.encoding)

--- a/jsonschema_spec/handlers/protocols.py
+++ b/jsonschema_spec/handlers/protocols.py
@@ -1,9 +1,7 @@
 from typing import Optional
 from typing import Protocol
-from typing import runtime_checkable
 
 
-@runtime_checkable
 class SupportsRead(Protocol):
     def read(self, amount: Optional[int] = 0) -> str:
         ...

--- a/jsonschema_spec/handlers/requests.py
+++ b/jsonschema_spec/handlers/requests.py
@@ -1,25 +1,31 @@
 """JSONSchema spec handlers requests module."""
 from contextlib import closing
 from io import StringIO
-from typing import Any
+from typing import ContextManager
+from typing import Optional
 
 import requests
 
 from jsonschema_spec.handlers.file import BaseFilePathHandler
+from jsonschema_spec.handlers.file import FileHandler
+from jsonschema_spec.handlers.protocols import SupportsRead
 
 
 class UrlRequestsHandler(BaseFilePathHandler):
     """URL (requests) scheme handler."""
 
-    def __init__(self, *allowed_schemes: str, **options: Any):
-        self.timeout = options.pop("timeout", 10)
-        super().__init__(**options)
-        self.allowed_schemes = list(allowed_schemes)
+    def __init__(
+        self,
+        *allowed_schemes: str,
+        file_handler: Optional[FileHandler] = None,
+        timeout: int = 10
+    ):
+        super().__init__(*allowed_schemes, file_handler=file_handler)
+        self.timeout = timeout
 
-    def _open(self, url: str) -> Any:
-        response = requests.get(url, timeout=self.timeout)
+    def _open(self, uri: str) -> ContextManager[SupportsRead]:
+        response = requests.get(uri, timeout=self.timeout)
         response.raise_for_status()
 
         data = StringIO(response.text)
-        with closing(data) as fh:
-            return super().__call__(fh)
+        return closing(data)

--- a/jsonschema_spec/handlers/urllib.py
+++ b/jsonschema_spec/handlers/urllib.py
@@ -1,18 +1,25 @@
 """JSONSchema spec handlers urllib module."""
-from typing import Any
+from contextlib import closing
+from typing import ContextManager
+from typing import Optional
 from urllib.request import urlopen
 
 from jsonschema_spec.handlers.file import BaseFilePathHandler
+from jsonschema_spec.handlers.file import FileHandler
+from jsonschema_spec.handlers.protocols import SupportsRead
 
 
 class UrllibHandler(BaseFilePathHandler):
     """URL (urllib) scheme handler."""
 
-    def __init__(self, *allowed_schemes: str, **options: Any):
-        self.timeout = options.pop("timeout", 10)
-        super().__init__(**options)
-        self.allowed_schemes = list(allowed_schemes)
+    def __init__(
+        self,
+        *allowed_schemes: str,
+        file_handler: Optional[FileHandler] = None,
+        timeout: int = 10
+    ):
+        super().__init__(*allowed_schemes, file_handler=file_handler)
+        self.timeout = timeout
 
-    def _open(self, url: str) -> Any:
-        with urlopen(url, timeout=self.timeout) as fh:
-            return super().__call__(fh)
+    def _open(self, uri: str) -> ContextManager[SupportsRead]:
+        return closing(urlopen(uri, timeout=self.timeout))

